aipmiddleware/validateparent.go

```diff
@@ -0,0 +1,89 @@
+package aipmiddleware
+
+import (
+	"context"
+
+	"go.einride.tech/aip/resourcename"
+	"google.golang.org/genproto/googleapis/api/annotations"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protoreflect"
+)
+
+// ParentValidator validates that response resources are children of requested parent resources.
+type ParentValidator struct {
+	// LogHook is called when parent validation fails, with the failing request/response, parent/name combination.
+	LogHook func(info *grpc.UnaryServerInfo, request, response proto.Message, parent, name string)
+}
+
+// UnaryServerInterceptor implements grpc.UnaryServerInterceptor.
+func (p *ParentValidator) UnaryServerInterceptor(
+	ctx context.Context,
+	request interface{},
+	info *grpc.UnaryServerInfo,
+	handler grpc.UnaryHandler,
+) (interface{}, error) {
+	protoRequest, ok := request.(proto.Message)
+	if !ok {
+		return nil, status.Error(codes.Internal, "parent validator got non-proto request")
+	}
+	protoReflectRequest := protoRequest.ProtoReflect()
+	// Request must have a parent field of string kind.
+	parentField := protoReflectRequest.Descriptor().Fields().ByName("parent")
+	if parentField == nil ||
+		parentField.IsMap() ||
+		parentField.IsList() ||
+		parentField.Kind() != protoreflect.StringKind {
+		return handler(ctx, request)
+	}
+	// The parent field must have a resource reference with a child type.
+	parentResourceReference := proto.GetExtension(
+		parentField.Options(), annotations.E_ResourceReference,
+	).(*annotations.ResourceReference)
+	if parentResourceReference.GetChildType() == "" {
+		return handler(ctx, request)
+	}
+	parent := protoReflectRequest.Get(parentField).String()
+	if parent == "" {
+		return handler(ctx, request)
+	}
+	response, err := handler(ctx, request)
+	if err != nil {
+		return nil, err
+	}
+	protoResponse, ok := response.(proto.Message)
+	if !ok {
+		return nil, status.Error(codes.Internal, "parent validator got non-proto response")
+	}
+	protoResponse.ProtoReflect().Range(func(field protoreflect.FieldDescriptor, value protoreflect.Value) bool {
+		if field.Kind() != protoreflect.MessageKind || !field.IsList() {
```

Comment:
maps are also messages, but protoreflect will panic if you call `.Message` on it.

Reply:
Yes, but since we're requiring `field.IsList()` here we should be good, right?

Reply:
Well yes, but only because you can't have repeated maps. I think we should explicit with !IsMap

Reply:
Because the maps and message is so surprising in protoreflect (and proto in general) I think always checking for maps is the safest option.

Reply:
I see what you mean, but I think we should go by the definition of `IsList` and `IsMap` as mutually exclusive:

```go
	// IsList reports whether this field represents a list,
	// where the value type for the associated field is a List.
	// It is equivalent to checking whether Cardinality is Repeated and
	// that IsMap reports false.
```

Reply:
I.e. if `IsList` is true, we know by definition that `IsMap` will be false.